src/com/amazon/ion/impl/IonReaderBinaryRawX.java

```diff
@@ -1181,81 +1202,106 @@ protected final Timestamp readTimestamp(int len) throws IOException
+                        utf8InputBuffer.position(),
+                        utf8InputBuffer.array(),
+                        0,
+                        carryoverBytes
+                );
             }
         }
+
         _local_remaining = save_limit;
-        return new String(chars, 0, ii);
+
+        decodingBuffer.flip();
+        return decodingBuffer.toString();
     }
-    private final int readUnicodeScalar() throws IOException
-    {
-        int c = -1, b;
-        b = read();
-        // ascii is all good, even -1 (eof)
-        if (IonUTF8.isOneByteUTF8(b)) {
-            return b;
+
+    private String readStringWithReusableBuffer(int numberOfBytes) throws IOException {
+        int save_limit = NO_LIMIT;
+        if (_local_remaining != NO_LIMIT) {
+            save_limit = _local_remaining - numberOfBytes;
         }
-        switch(IonUTF8.getUTF8LengthFromFirstByte(b)) {
-        case 2:
-            // now we start gluing the multi-byte value together
-            assert((b & 0xe0) == 0xc0);
-            // for values from 0x80 to 0x7FF (all legal)
-            int b2 = read();
-            if (!IonUTF8.isContinueByteUTF8(b2)) throwUTF8Exception();
-            c = IonUTF8.twoByteScalar(b, b2);
-            break;
-        case 3:
-            assert((b & 0xf0) == 0xe0);
-            // for values from 0x800 to 0xFFFFF (NOT all legal)
-            b2 = read();
-            if (!IonUTF8.isContinueByteUTF8(b2)) throwUTF8Exception();
-            int b3 = read();
-            if (!IonUTF8.isContinueByteUTF8(b3)) throwUTF8Exception();
-            c = IonUTF8.threeByteScalar(b, b2, b3);
-            break;
-        case 4:
-            assert((b & 0xf8) == 0xf0);
-            // for values from 0x010000 to 0x1FFFFF (NOT all legal)
-            b2 = read();
-            if (!IonUTF8.isContinueByteUTF8(b2)) throwUTF8Exception();
-            b3 = read();
-            if (!IonUTF8.isContinueByteUTF8(b3)) throwUTF8Exception();
-            int b4 = read();
-            if (!IonUTF8.isContinueByteUTF8(b4)) throwUTF8Exception();
-            c = IonUTF8.fourByteScalar(b, b2, b3, b4);
-            if (c > 0x10FFFF) {
-                throw new IonException("illegal utf value encountered in input utf-8 stream");
-            }
-            break;
-        default:
-            throwUTF8Exception();
+        _local_remaining = numberOfBytes;
+        readAll(utf8InputBuffer.array(), 0, numberOfBytes);
+        _local_remaining = save_limit;
+
+        utf8InputBuffer.position(0);
+        utf8InputBuffer.limit(numberOfBytes);
+
+        utf8DecodingBuffer.position(0);
+        utf8DecodingBuffer.limit(utf8DecodingBuffer.capacity());
+
+        utf8CharsetDecoder.reset();
+        CoderResult coderResult = utf8CharsetDecoder.decode(utf8InputBuffer, utf8DecodingBuffer, true);
+        if (coderResult.isError()) {
+            throw new IonException("Illegal value encountered while validating UTF-8 data in input stream.");
```

Comment:
let's add `coderResult.toString()` as is done above on line 1259

Reply:
Good call, done.